### PR TITLE
Issues 1198: Maven specific features support for @TestArtifact annotation

### LINKED ISSUE #1198: Maven classifiers support for @TestArtifact annotation
It should be possible to generate Maven specific artifacts with classifiers using `@TestAtrifact` annotation. This can be done generic way using generator strategy interface below:
```
public interface ArtifactGeneratorStrategy<T extends ArtifactGenerator>
{

    Path generateArtifact(T artifactGenerator, 
                          String id,
                          String version,
                          int size);

}
```
For Maven there will be `MavenArtifactWithClassifiersGeneratorStrategy` implementation, which will generate common classifiers like `sources`, `javadoc` etc. So the `@TestArtifact` annotation usage will looks like below:
```
@TestArtifact(repository = REPOSITORY_SNAPSHOTS, 
              id="org.carlspring.strongbox:test-snapshot", 
              versions = { "1.0-20180328.195810-1",
                           "1.0-20180328.195810-2",
                           "1.0-20180328.195810-3" }, 
              generator = MavenArtifactGenerator.class,
              strategy = MavenArtifactClassifiersGeneratorStrategy.class
)
```
To simplify the `@TestArtifact`  definition it should be also possible to define Layout specific shortcut meta annotations, so there will be `@MavenArtifactWithClassifiers` which is `@TestArtifact` with some predefined attributes:
```
@MavenArtifactWithClassifiers(repository = REPOSITORY_SNAPSHOTS, 
                              id = "org.carlspring.strongbox:test-snapshot", 
                              versions = { "1.0-20180328.195810-1",
                                           "1.0-20180328.195810-2",
                                           "1.0-20180328.195810-3" }
)
```
The above `@MavenArtifactWithClassifiers` annotation can be defined like below:
```
@Target(ElementType.PARAMETER)
@Retention(RetentionPolicy.RUNTIME)
@Documented
@TestArtifact(generator = MavenArtifactGenerator.class,
              strategy = MavenArtifactClassifiersGeneratorStrategy.class)
public @interface MavenArtifactWithClassifiers
 {

    /**
     * {@link Storage} ID.
     */
    @AliasFor(annotation = TestArtifact.class)
    String storage() default "storage0";

    /**
     * {@link Repository} ID.
     */
    @AliasFor(annotation = TestArtifact.class)
    String repository() default "";

    /**
     * The artifact "GA" (ex. "org.carlspring.test:test-artifact").
     */
    @AliasFor(annotation = TestArtifact.class)
    String id();
    
    /**
     * The {@link MavenArtifactCoordinates} versions.
     */
    @AliasFor(annotation = TestArtifact.class)
    String[] versions();
}
```

### PR DESCRIPTION
**Pull request by sbespalov**
Fixes #1198 
- `@MavenTestArtifact` annotation implemented
- `ArtifactMetadataServiceSnapshotsTest` reworked to use `@TestRespository` and `@TestArtifact`

### CODE REVIEW

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout/strongbox-storage-maven-layout-provider/src/test/java/org/carlspring/strongbox/services/ArtifactMetadataServiceSnapshotsTest.java [hunk 1]
```diff
@@ -177,56 +160,44 @@ public void testAddTimestampedSnapshotVersionToMetadata()
...
         assertNotNull(metadataAfter);
-        assertFalse(MetadataHelper.containsTimestampedSnapshotVersion(metadataAfter, deletedArtifact.getVersion()));
+        assertFalse(MetadataHelper.containsTimestampedSnapshotVersion(metadataAfter, coordinates.getVersion()));
     }
 
     @Test
-    public void testSnapshotWithoutTimestampMetadataRebuild()
+    @ExtendWith({ RepositoryManagementTestExecutionListener.class, ArtifactManagementTestExecutionListener.class })
+    public void testSnapshotWithoutTimestampMetadataRebuild(@MavenSnapshotRepository Repository repository,
+                                                            @MavenArtifactWithClassifiers(repository = REPOSITORY_SNAPSHOTS, id = "org.carlspring.strongbox.snapshots:metadata", versions = "2.0-SNAPSHOT") List<Path> snapshotArtifactPath)
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout/strongbox-storage-maven-layout-provider/src/test/java/org/carlspring/strongbox/services/ArtifactMetadataServiceSnapshotsTest.java]: So, how do you actually specify what classifiers to use when generating the artifacts? I'm not sure I get it.

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout/strongbox-storage-maven-layout-provider/src/test/java/org/carlspring/strongbox/services/ArtifactMetadataServiceSnapshotsTest.java [hunk 2]
```diff
@@ -36,45 +59,21 @@
...
-            throws Exception
-    {
-        createRepository(STORAGE0, REPOSITORY_SNAPSHOTS, RepositoryPolicyEnum.SNAPSHOT.getPolicy(), false);
-    }
-
     @Test
-    public void testSnapshotMetadataRebuild()
+    @ExtendWith({ RepositoryManagementTestExecutionListener.class, ArtifactManagementTestExecutionListener.class })
+    public void testSnapshotMetadataRebuild(@MavenSnapshotRepository Repository repository,
+                                            @MavenSnapshotArtifactWithClassifiers(id = "org.carlspring.strongbox:strongbox-metadata") List<Path> snapshotArtifacts)
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout/strongbox-storage-maven-layout-provider/src/test/java/org/carlspring/strongbox/services/ArtifactMetadataServiceSnapshotsTest.java]: I'm not sure I get the idea behind this `List<Path> ...`. I don't understand... 
Would you mind explaining this sourcery to me? :-o

Comment (sbespalov) [reply strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout/strongbox-storage-maven-layout-provider/src/test/java/org/carlspring/strongbox/services/ArtifactMetadataServiceSnapshotsTest.java]: it explained in #1165 under "To create multiple test artifact versions" section.

We have `@MavenSnapshotArtifactWithClassifiers` here which is shortcut for `@MavenArtifactWithClassifiers` with predefined 5 snapshot versions. So the `List<Path>` will contain 5 elements, each for every version.

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout/strongbox-storage-maven-layout-provider/src/test/java/org/carlspring/strongbox/services/ArtifactMetadataServiceSnapshotsTest.java (hunk 1 above)
Comment (sbespalov) [reply strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout/strongbox-storage-maven-layout-provider/src/test/java/org/carlspring/strongbox/services/ArtifactMetadataServiceSnapshotsTest.java]: Classifiers defined in `MavenArtifactWithClassifiersGeneratorStrategy.CLASSIFIERS`

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout/strongbox-storage-maven-layout-provider/src/test/java/org/carlspring/strongbox/services/ArtifactMetadataServiceSnapshotsTest.java (hunk 2 above)
Comment (sbespalov) [reply strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout/strongbox-storage-maven-layout-provider/src/test/java/org/carlspring/strongbox/services/ArtifactMetadataServiceSnapshotsTest.java]: If you have only one version then `List` not required and single `Path` can be injected.

**Comment by sbespalov** [comment]
@carlspring classifiers parameter implemented as you requested (see [here](https://github.com/strongbox/strongbox/pull/1199/files#diff-d58def3a919b56e55a36134fc5c5e122R313))

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout/strongbox-storage-maven-layout-provider/src/test/java/org/carlspring/strongbox/services/ArtifactMetadataServiceSnapshotsTest.java (hunk 1 above)
Comment (sbespalov) [reply strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout/strongbox-storage-maven-layout-provider/src/test/java/org/carlspring/strongbox/services/ArtifactMetadataServiceSnapshotsTest.java]: @carlspring classifiers parameter implemented as you requested (see [here](https://github.com/strongbox/strongbox/pull/1199/files#diff-d58def3a919b56e55a36134fc5c5e122R313))

Comment (carlspring) [reply strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout/strongbox-storage-maven-layout-provider/src/test/java/org/carlspring/strongbox/services/ArtifactMetadataServiceSnapshotsTest.java]: Thank you very much! :)

**Comment by carlspring** [comment]
Thanks, @sbespalov ! :)

**Pull request merged**